In [1]:
from torchvision import datasets as datasets_torch
from torchvision.transforms import ToTensor, Normalize, Compose
from torch.utils.data import DataLoader

## MNIST
### Load the data

In [39]:
dataset_name = 'MNIST'

transforms = []

transforms += [ToTensor()]

# Normalize
# transforms.append(Normalize(mean=(0.1307,), std=(0.3081,)))

transforms = Compose(transforms)

In [46]:
dataset = getattr(datasets_torch, dataset_name)

dataset_train = dataset(root='/tmp/{}'.format(dataset), train=True, download=True, transform=transforms)
# dataset_test = dataset(root='/tmp/{}'.format(dataset), train=False, download=True, transform=transforms)

data = DataLoader(dataset_train, batch_size=60000, num_workers=4, shuffle=True)


### Finding the mean and variance

In [47]:
for (X, Y) in data:
    print(X.shape, Y.shape)
    print(X.max(), X.min())
    print(X.mean(), X.std())
    break

torch.Size([60000, 1, 28, 28]) torch.Size([60000])
tensor(1.) tensor(0.)
tensor(0.1307) tensor(0.3081)


As seen above, the empirical mean of the dataset is **0.1307** and the empirical standard deviation is **0.3081**

## SVHN

In [2]:
dataset_name = 'SVHN'

transforms = []

transforms += [ToTensor()]

# Normalize
# transforms.append(Normalize(mean=(0.1307,), std=(0.3081,)))

transforms = Compose(transforms)

In [3]:
dataset = getattr(datasets_torch, dataset_name)

dataset_train = dataset(root='/tmp/{}'.format(dataset_name), download=True, transform=transforms)

data = DataLoader(dataset_train, batch_size=80000, num_workers=4, shuffle=True)


Using downloaded and verified file: /tmp/SVHN/train_32x32.mat


### Finding the mean and variance

In [4]:
for (X, Y) in data:
    print(X.shape, Y.shape)
    print(X.max(), X.min())
    print(X.mean(dim=(0, 2, 3)), X.std(dim=(0, 2, 3,)))
    break

torch.Size([73257, 3, 32, 32]) torch.Size([73257])
tensor(1.) tensor(0.)
tensor([0.4377, 0.4438, 0.4728]) tensor([0.1980, 0.2010, 0.1970])


As seen above, the empirical mean of the dataset is **(0.4377, 0.4438, 0.4728)** and the empirical standard deviation is **(0.1980, 0.2010, 0.1970)**